contracts/axelarnet-gateway/src/contract/execute.rs

```diff
@@ -0,0 +1,134 @@
+use axelar_wasm_std::nonempty;
+use cosmwasm_std::{to_json_binary, Addr, Api, HexBinary, Response, Storage, WasmMsg};
+use error_stack::{report, Result, ResultExt};
+use router_api::client::Router;
+use router_api::{Address, ChainName, CrossChainId, Message};
+use sha3::{Digest, Keccak256};
+
+use crate::contract::Error;
+use crate::events::AxelarnetGatewayEvent;
+use crate::msg::AxelarExecutableExecuteMsg;
+use crate::state::{self};
+
+// TODO: Retrieve the actual tx hash from core, since cosmwasm doesn't provide it. Use a placeholder in the meantime.
+const PLACEHOLDER_TX_HASH: &str =
+    "0x0000000000000000000000000000000000000000000000000000000000000000";
+
+pub(crate) fn call_contract(
+    store: &mut dyn Storage,
+    router: &Router,
+    chain_name: ChainName,
+    sender: Addr,
+    destination_chain: ChainName,
+    destination_address: Address,
+    payload: HexBinary,
+) -> Result<Response, Error> {
+    let counter =
+        state::increment_message_counter(store).change_context(Error::InvalidStoreAccess)?;
+
+    let cc_id = CrossChainId {
+        source_chain: chain_name.into(),
+        message_id: nonempty::String::try_from(format!("{0}-{1}", PLACEHOLDER_TX_HASH, counter))
+            .change_context(Error::MessageIdConstructionFailed)?,
+    };
+
+    let payload_hash = Keccak256::digest(payload.as_slice()).into();
+
+    let msg = Message {
+        cc_id: cc_id.clone(),
+        source_address: Address::try_from(sender.clone().into_string())
+            .change_context(Error::InvalidSender(sender))?,
+        destination_chain,
+        destination_address,
+        payload_hash,
+    };
+
+    state::save_incoming_msg(store, cc_id, &msg).change_context(Error::InvalidStoreAccess)?;
+
+    let wasm_msg = router
+        .route(vec![msg.clone()])
+        .ok_or(Error::RoutingFailed)?;
+
+    Ok(Response::new()
+        .add_message(wasm_msg)
+        .add_event(AxelarnetGatewayEvent::ContractCalled { msg, payload }.into()))
+}
+
+// because the messages came from the router, we can assume they are already verified
+pub(crate) fn route_outgoing_messages(
+    store: &mut dyn Storage,
+    msgs: Vec<Message>,
+) -> Result<Response, Error> {
+    for msg in msgs.iter() {
+        state::save_outgoing_message(store, msg.cc_id.clone(), msg.clone())
+            .change_context(Error::SaveOutgoingMessage)?;
+    }
+
+    Ok(Response::new().add_events(
+        msgs.into_iter()
+            .map(|msg| AxelarnetGatewayEvent::Routing { msg }.into()),
+    ))
+}
+
+pub(crate) fn route_incoming_messages(
```

Comment:
same comment about nomenclature here, I find these names confusing

Reply:
we should change in the gateway as well. Since the frame of reference is confusing, we can try to be more explicit. Perhaps something like `route_from_router`, `router_to_gateway`, `route_to_chain`